crates/torii/grpc/src/server/subscriptions/event.rs

```diff
@@ -0,0 +1,134 @@
+use std::sync::Arc;
+use std::task::{Context, Poll};
+
+use futures::Stream;
+use futures_util::StreamExt;
+use rand::Rng;
+use starknet_crypto::FieldElement;
+use tokio::sync::mpsc::{channel, Receiver, Sender};
+use tokio::sync::RwLock;
+use torii_core::error::{Error, ParseError};
+use torii_core::simple_broker::SimpleBroker;
+use torii_core::sql::FELT_DELIMITER;
+use torii_core::types::Event;
+use tracing::{error, trace};
+
+use crate::proto;
+use crate::proto::world::SubscribeEventsResponse;
+
+pub(crate) const LOG_TARGET: &str = "torii::grpc::server::subscriptions::event";
+
+pub struct EventSubscriber {
+    /// Event keys that the subscriber is interested in
+    keys: Vec<FieldElement>,
+    /// The channel to send the response back to the subscriber.
+    sender: Sender<Result<proto::world::SubscribeEventsResponse, tonic::Status>>,
+}
+
+#[derive(Default)]
+pub struct EventManager {
+    subscribers: RwLock<HashMap<usize, EventSubscriber>>,
+}
+
+impl EventManager {
+    pub async fn add_subscriber(
+        &self,
+        keys: Vec<FieldElement>,
+    ) -> Result<Receiver<Result<proto::world::SubscribeEventsResponse, tonic::Status>>, Error> {
+        let id = rand::thread_rng().gen::<usize>();
+        let (sender, receiver) = channel(1);
+
+        // NOTE: unlock issue with firefox/safari
+        // initially send empty stream message to return from
+        // initial subscribe call
+        let _ = sender.send(Ok(SubscribeEventsResponse { event: None })).await;
+
+        self.subscribers.write().await.insert(id, EventSubscriber { keys, sender });
+
+        Ok(receiver)
+    }
+
+    pub(super) async fn remove_subscriber(&self, id: usize) {
+        self.subscribers.write().await.remove(&id);
+    }
+}
+
+#[must_use = "Service does nothing unless polled"]
+pub struct Service {
+    subs_manager: Arc<EventManager>,
+    simple_broker: Pin<Box<dyn Stream<Item = Event> + Send>>,
+}
+
+impl Service {
+    pub fn new(subs_manager: Arc<EventManager>) -> Self {
+        Self { subs_manager, simple_broker: Box::pin(SimpleBroker::<Event>::subscribe()) }
+    }
+
+    async fn publish_updates(subs: Arc<EventManager>, event: &Event) -> Result<(), Error> {
+        let mut closed_stream = Vec::new();
+        let keys = event
+            .keys
+            .trim_end_matches(FELT_DELIMITER)
+            .split(FELT_DELIMITER)
+            .map(FieldElement::from_str)
+            .collect::<Result<Vec<_>, _>>()
+            .map_err(ParseError::from)?;
+        let data = event
+            .data
+            .trim_end_matches(FELT_DELIMITER)
+            .split(FELT_DELIMITER)
+            .map(FieldElement::from_str)
+            .collect::<Result<Vec<_>, _>>()
+            .map_err(ParseError::from)?;
+
+        for (idx, sub) in subs.subscribers.read().await.iter() {
+            // publish all updates if ids is empty or only ids that are subscribed to
+            if sub.keys.is_empty() || keys.starts_with(&sub.keys) {
```

Comment:
I guess that's the most common operation on regular event, we just want a sub-array from the beginning. We may update this later if needed.

We may document the behavior though, to ensure the user can understand the clause. I'll add an issue on the dojo-book to ensure that we describe better what's available and how. :)

Thanks for the work done here mate!